// @lucia-auth/oauth@2.0.1/dist/providers/osu.js
import { createUrl, handleRequest, authorizationHeaders } from "../request.js";
import { providerUserAuth } from "../core.js";
import { scope, generateState } from "../utils.js";
const PROVIDER_ID = "osu";
export const osu = (auth, config) => {
    const getOsuTokens = async (code) => {
        const request = new Request("https://osu.ppy.sh/oauth/token", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded"
            },
            body: new URLSearchParams({
                client_id: config.clientId,
                client_secret: config.clientSecret,
                grant_type: "authorization_code",
                redirect_uri: config.redirectUri,
                code
            }).toString()
        });
        const tokens = await handleRequest(request);
        return {
            accessToken: tokens.access_token,
            refreshToken: tokens.refresh_token,
            accessTokenExpiresIn: tokens.expires_in
        };
    };
    const getOsuUser = async (accessToken) => {
        const request = new Request("https://osu.ppy.sh/api/v2/me/osu", {
            headers: authorizationHeaders("bearer", accessToken)
        });
        const osuUser = await handleRequest(request);
        return osuUser;
    };
    return {
        getAuthorizationUrl: async () => {
            const state = generateState();
            const url = createUrl("https://osu.ppy.sh/oauth/authorize", {
                response_type: "code",
                client_id: config.clientId,
                scope: scope(["identify"], config.scope),
                redirect_uri: config.redirectUri,
                state
            });
            return [url, state];
        },
        validateCallback: async (code) => {
            const osuTokens = await getOsuTokens(code);
            const osuUser = await getOsuUser(osuTokens.accessToken);
            const providerUserId = osuUser.id.toString();
            const osuUserAuth = await providerUserAuth(auth, PROVIDER_ID, providerUserId);
            return {
                ...osuUserAuth,
                osuUser,
                osuTokens
            };
        }
    };
};
