// @lucia-auth/oauth@2.0.1/dist/providers/patreon.js
import { createUrl, handleRequest, authorizationHeaders } from "../request.js";
import { providerUserAuth } from "../core.js";
import { scope, generateState } from "../utils.js";
const PROVIDER_ID = "patreon";
export const patreon = (auth, config) => {
    const getPatreonTokens = async (code) => {
        const requestUrl = createUrl("https://www.patreon.com/api/oauth2/token", {
            client_id: config.clientId,
            client_secret: config.clientSecret,
            code,
            grant_type: "authorization_code",
            redirect_uri: config.redirectUri
        });
        const request = new Request(requestUrl, {
            method: "POST"
        });
        const tokens = await handleRequest(request);
        return {
            accessToken: tokens.access_token,
            refreshToken: tokens.refresh_token ?? null,
            accessTokenExpiresIn: tokens.expires_in
        };
    };
    const getPatreonUser = async (accessToken) => {
        const requestUrl = createUrl("https://www.patreon.com/api/oauth2/v2/identity", {
            "fields[user]": "about,email,full_name,hide_pledges,image_url,is_email_verified,url"
        });
        const request = new Request(requestUrl, {
            headers: authorizationHeaders("bearer", accessToken)
        });
        const { data: patreonUser } = await handleRequest(request);
        return patreonUser;
    };
    return {
        getAuthorizationUrl: async () => {
            const state = generateState();
            const url = createUrl("https://www.patreon.com/oauth2/authorize", {
                client_id: config.clientId,
                redirect_uri: config.redirectUri,
                scope: scope(["identity"], config.scope),
                response_type: "code",
                state
            });
            return [url, state];
        },
        validateCallback: async (code) => {
            const patreonTokens = await getPatreonTokens(code);
            const patreonUser = await getPatreonUser(patreonTokens.accessToken);
            const providerUserId = patreonUser.id;
            const patreonUserAuth = await providerUserAuth(auth, PROVIDER_ID, providerUserId);
            return {
                ...patreonUserAuth,
                patreonUser,
                patreonTokens
            };
        }
    };
};
